### DISCUSSION TITLE
Custom Blocks for Custom Models

### DISCUSSION THREAD

**Discussion by kohlerpop1** [PurpurMC/Purpur#1035] [__Dismissed__] [answered]
Lots of people including myself are currently using string, mushroom blocks, and many other data variable blocks in order to have custom textures ingame. 

I think it would be amazing to have custom blocks, almost identical to the barrier block but with changable data. 

For instance, when using string we have to mess with lots of block states and the same for mushroom. This is a massive pain for lots of developers, 3D modelers, and many more. A simple 2 new blocks could make a enormous difference. I think all blocks should be non-occluding in nature as this would allow for smaller than 16x16 or 32x32 pixel square blocks to not have xray glitching.

I think the following blocks would be amazing to have.
The small list below is an example of the block with similar properties.
- Non-Passible-Block:  Ex: STONE_BLOCK
- Passible-Block:  Ex: COBWEB, but without reduced speed.

Block List:
1. Non-Passible-Block, with SHAPE (0,0,0,16,16,16) and IntegerProperty of 'DATA' from 1-1000 with Default = 0.
   - This would allow for custom "normal" blocks, like stone, to be placed/used around the world as developers and server owners see fit. 
   - We would be able to configure this specific block into json files to render from another full block model with this.
2. Passible-Block, with SHAPE (0,0,0,16,16,16) and IntegerProperty of 'DATA' up from 1-1000 with Default = 0.
   - This would allow for custom "passable" blocks, like a bush, flower or cobweb, to be placed/used around the world as well. 
   - This is almost identical to concept # 1, except for naturally passable.

Ideally, there could also be 2 more custom blocks exactly as the ones above but with a smaller model/shape but they are not as big of a necessity as the first two. Please definitely consider it. It does not look like it would be too hard to add but I currently do not have the adequate time or exact knowledge on how to do this. I have looked through the source code and thought of doing a patch myself but figured I would leave it the the experts. 

The main takeaways from this is that the custom blocks need to be normal in shape, non-occluding to prevent xray glitching, and have lots of changeable data available (preferrably on 1 property as well).

This is mainly to prevent stuff like [this](https://paste.gg/p/anonymous/b03daaa1eb1242d1aac13c5c3cd8fbef) from happening!

Thanks so much!

**Comment by BillyGalbreath** [comment, marked as answer]
This isn't even remotely possible.

We can't just add new blocks to the game, the same way you can't just add new blocks to the game. The client has to know of these new blocks. That's why you have to use predicates on existing blocks. It's the only way unless you step over into Fabric/Forge modding.
